Source/Core/Common/Network.cpp

```diff
@@ -72,4 +81,91 @@ std::optional<MACAddress> StringToMacAddress(std::string_view mac_string)
 
   return std::make_optional(mac);
 }
+
+EthernetHeader::EthernetHeader() = default;
+
+EthernetHeader::EthernetHeader(u16 ether_type)
+{
+  ethertype = htons(ether_type);
+}
+
+u16 EthernetHeader::Size() const
+{
+  return static_cast<u16>(SIZE);
+}
+
+IPv4Header::IPv4Header() = default;
+
+IPv4Header::IPv4Header(u16 data_size, u8 ip_proto, const sockaddr_in& from, const sockaddr_in& to)
+{
+  version_ihl = 0x45;
+  total_len = htons(Size() + data_size);
+  flags_fragment_offset = htons(0x4000);
+  ttl = 0x40;
+  protocol = ip_proto;
+  std::memcpy(&source_addr, &from.sin_addr, IPV4_ADDR_LEN);
+  std::memcpy(&destination_addr, &to.sin_addr, IPV4_ADDR_LEN);
+
+  // Compute checksum
+  u32 checksum = 0;
+  u16 word = 0;
+  std::size_t index = 0;
+  const std::string_view ip_header{reinterpret_cast<char*>(this), Size()};
+  for (u8 b : ip_header)
+  {
+    if (index % 2 == 0)
+      word = b;
+    else
+      checksum += (word << 8) + b;
+    index += 1;
+  }
+  header_checksum = htons((checksum + (checksum >> 16) & 0xFFFF) ^ 0xFFFF);
```

Comment:
Shouldn't this be a loop? Something like this:

```cpp
while (checksum > 0xFFFF)
  checksum = (checksum >> 16) + (checksum & 0xFFFF); 

header_checksum = htons(checksum ^ 0xFFFF);
```

If you only perform one iteration of the loop, you may still end up with a checksum that is larger than 0xFFFF.

Reply:
Good catch. fixed.